irohad/ametsuchi/impl/postgres_query_executor.cpp

```diff
@@ -151,64 +213,78 @@ namespace iroha {
               deserialized_block)
               .value;
 
-      boost::transform(
-          range_gen(boost::size(block->transactions()))
-              | boost::adaptors::transformed(
-                    [&block](auto i) -> decltype(auto) {
-                      return block->transactions()[i];
-                    })
-              | boost::adaptors::filtered(pred),
-          std::back_inserter(result),
-          [&](const auto &tx) {
-            // TODO 03.10.18 andrei: IR-1729 Integrate query response factory
-            return *static_cast<shared_model::proto::Transaction *>(
-                clone(tx).get());
-          });
+      auto filtered_txs = range_gen(boost::size(block->transactions()))
+          | boost::adaptors::transformed([&block](auto i) -> decltype(auto) {
+                            return block->transactions()[i];
+                          })
+          | boost::adaptors::filtered(pred);
+      boost::transform(filtered_txs,
+                       std::back_inserter(result),
+                       [&](const auto &tx) { return clone(tx); });
 
       return result;
     }
 
-    template <typename Q, typename P, typename F, typename B>
-    QueryResponseBuilderDone PostgresQueryExecutorVisitor::executeQuery(F &&f,
-                                                                        B &&b) {
-      using T = concat<Q, P>;
+    template <typename QueryTuple,
+              typename PermissionTuple,
+              typename QueryExecutor,
+              typename ResponseCreator,
+              typename ErrResponse>
+    QueryExecutorResult PostgresQueryExecutorVisitor::executeQuery(
+        QueryExecutor &&query_executor,
+        ResponseCreator &&response_creator,
+        ErrResponse &&err_response) {
+      using T = concat<QueryTuple, PermissionTuple>;
       try {
-        soci::rowset<T> st = std::forward<F>(f)();
+        soci::rowset<T> st = std::forward<QueryExecutor>(query_executor)();
         auto range = boost::make_iterator_range(st.begin(), st.end());
 
         return apply(
-            viewPermissions<P>(range.front()), [range, &b](auto... perms) {
+            viewPermissions<PermissionTuple>(range.front()),
+            [this,
+             range,
+             &response_creator,
+             err_response =
+                 std::forward<ErrResponse>(err_response)](auto... perms) {
               bool temp[] = {not perms...};
               if (std::all_of(std::begin(temp), std::end(temp), [](auto b) {
                     return b;
                   })) {
-                return stateful_failed;
+                return logAndReturnErrorResponse(
+                    QueryErrorType::kStatefulFailed,
+                    err_response(),
+                    this->query_hash_,
```

Comment:
You pass three parameters which are class members. Maybe make logAndReturnErrorResponse a member of executor, so you do not need to pass parameters each time?

Reply:
The function is used in both `PostgresQueryExecutorVisitor` and `PostgresQueryExecutor`, thus making it member will force having two identical function definitions